MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermark.cs

```diff
@@ -0,0 +1,504 @@
+                    return true;
+                if (CheckNearestPoints(map, x, y - i, value))
+                    return true;
+
+            }
+            return false;
+        }
+
+        private bool CheckNearestPoints(bool[,] map, int x, int y, int value)
+        {
+            if (x < 0 || x >= _options.Extent || y < 0 || y >= _options.Extent)
+                return false;
+
+            if (x + 1 < _options.Extent)
+                if (Convert.ToInt32(map[x + 1, y]) != value)
+                    return true;
+
+            if (x - 1 >= 0)
+                if (Convert.ToInt32(map[x - 1, y]) != value)
+                    return true;
+
+            if (y + 1 < _options.Extent)
+                if (Convert.ToInt32(map[x, y + 1]) != value)
+                    return true;
+
+            if (y - 1 >= 0)
+                if (Convert.ToInt32(map[x, y - 1]) != value)
+                    return true;
+
+            return false;
+        }
+
+        private double Statistics(VectorTile tile, Polygon polygon, bool[,] map, Envelope envelopeTile, double extentDist, out int s0, out int s1)
+        {
+            s0 = 0;
+            s1 = 0;
+
+            foreach (var layer in tile.Layers)
+                foreach (var feature in layer.Features)
+                {
+                    var geometry = feature.Geometry;
+                    var coordinates = geometry.Coordinates;
+                    foreach (var coordinate in coordinates)
+                    {
+                        var coordinateMeters = CoordinateConverter.DegreesToMeters(coordinate);
+                        if (polygon.Contains(new Point(coordinateMeters)))
+                        {
+                            var x = Convert.ToInt32((coordinateMeters.X - envelopeTile.MinX) / extentDist);
+                            var y = Convert.ToInt32((coordinateMeters.Y - envelopeTile.MinY) / extentDist);
+                            if (x == _options.Extent || y == _options.Extent)
+                                continue;
+                            var mapValue = Convert.ToInt32(map[x, y]);
+
+                            if (mapValue == 1)
+                                s1++;
+                            else
+                                s0++;
+
+                        }
+                    }
+                }
+
+            if ((s0 == 0 && s1 == 0) || s0 + s1 < _options.CountPoints)
+                return -1;
+
+            return (double)Math.Abs(s0 - s1) / (s1 + s0);
+        }
+
+        private struct IntPoint
+        {
+            public int x;
+            public int y;
+            public IntPoint(int x, int y) { this.x = x; this.y = y; }
+        }
+
+        private List<IntPoint> GetOppositePoint(bool[,] map, int x, int y, int value)
+        {
+            int xRes;
+            int yRes;
+
+            var listPoints = new List<IntPoint>();
+
+            if (x + 1 < _options.Extent)
+                if (Convert.ToInt32(map[x + 1, y]) != value)
```

Comment:
There are a lot of conversions in the code.
Maybe it's better to make "map" variable a matrix of int's? or vice versa make "value" variable a bool?
It's the same in a bunch of other methods